agent/src/main/java/com/code_intelligence/jazzer/runtime/CoverageMap.java

```diff
@@ -21,18 +21,56 @@
 
 /**
  * Represents the Java view on a libFuzzer 8 bit counter coverage map.
- * By using a direct ByteBuffer, the counter array is shared directly with
- * native code.
+ * By using a direct ByteBuffer, the counters are shared directly with native
+ * code.
  */
 final public class CoverageMap {
-  public static ByteBuffer counters = ByteBuffer.allocateDirect(0);
+  private static final int MAX_NUM_COUNTERS = 1 << 20;
+
+  /**
+   * The collection of coverage counters directly interacted with by classes
+   * that are instrumented for coverage.
+   * The instrumentation assumes that this is always one contiguous block of
+   * memory, so it is allocated once at maximum size. Using a larger number here
+   * increases the memory usage of all fuzz targets, but has otherwise no impact
+   * on performance.
+   */
+  public static final ByteBuffer counters = ByteBuffer.allocateDirect(MAX_NUM_COUNTERS);
+
+  static {
+    initialize(counters);
+  }
+
+  private static final int INITIAL_NUM_COUNTERS = 1 << 9;
+
+  static {
+    registerNewCounters(0, INITIAL_NUM_COUNTERS);
+  }
+
+  /**
+   * The number of coverage counters that are currently registered with
+   * libFuzzer. This number grows dynamically as classes are instrumented and
+   * should be kept as low as possible as libFuzzer has to iterate over the
+   * whole map for every execution.
+   */
+  private static int currentNumCounters = INITIAL_NUM_COUNTERS;
 
   // Called via reflection.
   @SuppressWarnings("unused")
   public static void enlargeIfNeeded(int nextId) {
-    if (nextId >= counters.capacity()) {
-      registerNewCoverageCounters();
-      System.out.println("INFO: New number of inline 8-bit counters: " + counters.capacity());
+    int newNumCounters = currentNumCounters;
+    while (nextId >= newNumCounters) {
+      newNumCounters = 2 * currentNumCounters;
```

Comment:
Since we are always doubling the size of the counters seen by libFuzzer, we might increase the size from `1 << 19` to `1 << 20` even if only a single new counter is needed. I was wondering whether this might cause an overhead that we can reduce with other strategies to increase the size map. My thought here was since we know the size of the current 8-bit counters we registered and the maximum edge id we instrumented, we can check whether we on average allocate much more counters than we need.

Reply:
Determining the correct factor would require weighing the cost of additional counter blocks vs. larger counter blocks. We'll keep it on our (imaginary) list of things to benchmark.